org/mozilla/jss/ssl/javax/JSSEngineReferenceImpl.java

```diff
@@ -0,0 +1,1279 @@
+            s_socket.setReceiveBufferSize(BUFFER_SIZE);
+            s_socket.setSendBufferSize(BUFFER_SIZE);
+
+            s_ostream = s_socket.getOutputStream();
+
+            c_ostream = c_socket.getOutputStream();
+        } catch (Exception e) {
+            throw new RuntimeException("Unable to enable debug socket logging! " + e.getMessage(), e);
+        }
+    }
+
+    public void beginHandshake() {
+        debug("JSSEngine: beginHandshake()");
+
+        // We assume beginHandshake(...) is the entry point for initializing
+        // the buffer. In particular, wrap(...) / unwrap(...) *MUST* call
+        // beginHandshake(...) if ssl_fd == null.
+
+        // ssl_fd == null <-> we've not initialized anything yet.
+        if (ssl_fd == null) {
+            init();
+        }
+
+        // Always, reset the handshake status, using the existing
+        // socket and configuration (which might've been just created).
+        //
+        // Also, NSS's SSL_ResetHandshake takes a parameter "asServer" which
+        // is the opposite of "is_client".
+        if (SSL.ResetHandshake(ssl_fd, !is_client) == SSL.SECFailure) {
+            throw new RuntimeException("Unable to begin handshake: " + errorText(PR.GetError()));
+        }
+    }
+
+    public void closeInbound() {
+        debug("JSSEngine: closeInbound()");
+
+        PR.Shutdown(ssl_fd, PR.SHUTDOWN_RCV);
+        is_inbound_closed = true;
+    }
+
+    public void closeOutbound() {
+        debug("JSSEngine: closeOutbound()");
+
+        PR.Shutdown(ssl_fd, PR.SHUTDOWN_SEND);
+        is_outbound_closed = true;
+    }
+
+    public String getHostname() {
+        return hostname;
+    }
+
+    public Runnable getDelegatedTask() {
+        debug("JSSEngine: getDelegatedTask()");
+
+        // We fake being a non-blocking SSLEngine. In particular, we never
+        // export tasks as delegated tasks (e.g., OCSP checking), so this
+        // method will always return null.
+
+        return null;
+    }
+
+    private void queryEnabledCipherSuites() {
+        debug("JSSEngine: queryEnabledCipherSuites()");
+        ArrayList<SSLCipher> enabledCiphers = new ArrayList<SSLCipher>();
+
+        for (SSLCipher cipher : SSLCipher.values()) {
+            try {
+                if (SSL.CipherPrefGet(ssl_fd, cipher.getID())) {
+                    enabledCiphers.add(cipher);
+                }
+            } catch (Exception e) {
+                // Do nothing -- this shouldn't happen as SSLCipher should be
+                // synced with NSS. However, we'll just log this exception as
+                // a warning. At worst we fail to report that a cipher suite is
+                // enabled.
+                warn("Unable to get the value of cipher: " + cipher.name() + " (" + cipher.getID() + "): " + e.getMessage());
+            }
+        }
+
+        enabled_ciphers = enabledCiphers.toArray(new SSLCipher[0]);
+    }
+
+    public String[] getEnabledCipherSuites() {
```

Comment:
@edewata Should we move these to JSSEngine as well? These likely have the same implementation regardless of the implementation of `wrap` / `unwrap`...

Reply:
What's the purpose of having a separate JSSEngine and JSSEngineReferenceImpl? Do we plan to support other custom implementations? I'm thinking we can just have a single implementation in JSSEngine, then let people extend it and override the methods.

Reply:
Partly to split the implementation into the "core" portion and all the fluff along the side (cipher suite handling, ...) but also because we do plan on having two implementations: 
 
- `JSSEngineReferenceImpl`
- `JSSEngineOptimizedImpl`

The latter will have less debug information but should have much better performance. The issue with the current implementation is that it is likely really slow. Each `wrap`/`unwrap` operation incurs _several_ round-trips over the JNI layer, which really impacts performance. We can write an optimized implementation that reduces the number of round-trips over the JNI layer, at the expense of larger C methods and less debugging info.

This is fine, but I don't necessarily want to do it until we've measured the performance and realized we needed it.